src/Endpoints/class-graphql-metadata.php

```diff
@@ -0,0 +1,262 @@
+<?php
+/**
+ * GraphQL support class.
+ * Note: this class will only work if the WPGraphQL plugin is installed.
+ *
+ * @package Parsely
+ * @since 3.2.0
+ */
+
+declare(strict_types=1);
+
+namespace Parsely\Endpoints;
+
+use WP_Post;
+
+/**
+ * Injects Parse.ly Metadata to the GraphQL outputs
+ *
+ * @since 3.2.0
+ */
+class GraphQL_Metadata extends Metadata_Endpoint {
+	private const GRAPHQL_VERSION          = '1.0.0';
+	private const GRAPHQL_CONTAINER_TYPE   = 'ParselyMetaContainer';
+	private const GRAPHQL_AUTHOR_TYPE      = 'ParselyAuthor';
+	private const GRAPHQL_MAIN_ENTITY_TYPE = 'ParselyMainEntityOfPage';
+	private const GRAPHQL_IMAGE_TYPE       = 'ParselyImage';
+	private const GRAPHQL_PUBLISHER_TYPE   = 'ParselyPublisher';
+	private const GRAPHQL_META_TYPE        = 'ParselyMeta';
+
+	/**
+	 * Register fields in WPGraphQL plugin
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return void
+	 */
+	public function run(): void {
+		if ( class_exists( 'WPGraphQL' ) ) {
```

Comment:
I'd like to see this check moved outside of this class. Wrap it around the call to `run()` instead.

Since it's protecting a call to an action that only appears in WPGraphQL, then it may not even be needed anyway, since the action would never run if the WPGraphQL isn't active anyway.

Reply:
> I'd like to see this check moved outside of this class. Wrap it around the call to run() instead.

Would this mean putting moving the check in `wp-parsely.php`? I think it's good that we have as least logic as possible in that file, so I'm not sure where else to put it.

> Since it's protecting a call to an action that only appears in WPGraphQL, then it may not even be needed anyway, since the action would never run if the WPGraphQL isn't active anyway.

Does it have an impact on performance/memory usage to enqueue too many things? I see it as a waste of resources to enqueue a function for all `wp-parsely` users, most of which won't be using GraphQL. Doing just one conditional is more performant than an `add_action`.

Reply:
> Would this mean putting moving the check in wp-parsely.php?

Yes, but that's a good thing.

>  I think it's good that we have as least logic as possible in that file, so I'm not sure where else to put it.

That's flawed thinking IMO. The main plugin file is the perfect place to be the [_composition root_](https://freecontent.manning.com/dependency-injection-in-net-2nd-edition-understanding-the-composition-root/) where classes and integrations are bootstrapped and started (or not).

> Does it have an impact on performance/memory usage to enqueue too many things? I see it as a waste of resources to enqueue a function for all wp-parsely users, most of which won't be using GraphQL. Doing just one conditional is more performant than an add_action.

That's fair 👍🏻